src/main/java/com/segment/analytics/android/integrations/firebase/FirebaseIntegration.java

```diff
@@ -3,88 +3,242 @@
-    super.onActivityStarted(activity);
+  @Override
+  public void identify(IdentifyPayload identify) {
+    super.identify(identify);
 
+    if (!isNullOrEmpty(identify.userId())) {
+      firebaseAnalytics.setUserId(identify.userId());
+    }
+    Map<String, Object> traits = identify.traits();
+    for (Map.Entry<String, Object> entry : traits.entrySet()) {
+      String trait = entry.getKey();
+      trait = trait.trim().replaceAll(" ", "_");
+      if (trait.length() > 40) {
+        trait = trimKey(trait);
+      }
+      String formattedValue;
+      if (entry.getValue() instanceof Date) {
+        Date value = (Date) entry.getValue();
+        formattedValue = formatDate(value);
+      } else {
+        formattedValue = String.valueOf(entry.getValue());
+      }
+      firebaseAnalytics.setUserProperty(trait, formattedValue);
+      logger.verbose("firebaseAnalytics.setUserProperty(%s, %s);", trait, formattedValue);
+    }
   }
 
-  @Override public void onActivityStopped(Activity activity) {
-    super.onActivityStopped(activity);
+  @Override
+  public void track(TrackPayload track) {
+    super.track(track);
 
+    String event = track.event();
+    String eventName = mapEvent(event);
+    Properties properties = track.properties();
+    Bundle formattedProperties = formatProperties(properties);
+    firebaseAnalytics.logEvent(eventName, formattedProperties);
+    logger.verbose("firebaseAnalytics.logEvent(%s, %s);", eventName, formattedProperties);
   }
 
-  @Override public void identify(IdentifyPayload identify) {
-    super.identify(identify);
-
-    if (!isNullOrEmpty(identify.userId())) {
-      mFirebaseAnalytics.setUserId(identify.userId());
+  private String mapEvent(String event) {
+    String eventName = event;
+    if (eventMapper.containsKey(eventName)) {
+      eventName = eventMapper.get(eventName);
     }
+    eventName = eventName.trim().replaceAll(" ", "_");
+    if (eventName.length() > 40) {
+      eventName = trimKey(eventName);
+    }
+    return eventName;
+  }
 
-    // mFirebaseAnalytics.setUserProperty(key, value);
-
+  private Bundle formatProperties(Properties properties) {
+    Bundle bundle = new Bundle();
+    if ((properties.revenue() != 0 || properties.total() != 0)
+        && isNullOrEmpty(properties.currency())) {
+      bundle.putString(Param.CURRENCY, "USD");
+    }
+    for (Map.Entry<String, Object> entry : properties.entrySet()) {
+      String property = entry.getKey();
+      property = mapProperty(property);
+      if (entry.getValue() instanceof Integer) {
+        int value = (int) entry.getValue();
+        bundle.putInt(property, value);
+        logger.verbose("bundle.putInt(%s, %s);", property, value);
+        continue;
+      }
+      if (entry.getValue() instanceof Double) {
+        double value = (double) entry.getValue();
+        bundle.putDouble(property, value);
+        logger.verbose("bundle.putDouble(%s, %s);", property, value);
```

Comment:
These don't need to be logged. The `logger.verbose("firebaseAnalytics.setUserProperty(%s, %s);", trait, formattedValue);` will log the key: value pairs already.

Reply:
I like that logging `bundle.put*` methods reminds users the data type of each value they're passing, for example: https://www.dropbox.com/s/jtkao2j3j6uxi5t/Screenshot%202017-09-07%2009.35.45.png?dl=0

You're right that the logger will otherwise display all key-value pairs passed in the FirebaseAnalytics.logEvent method.

What do you think?

Reply:
I think it's too verbose and inconsistent.

1. Too many log messages are distracting for customers. Currently this implementation will produce multiple lines of logs for each just for this one integration.
2. A Bundle is just a map container. We don't this for other containers, e.g. java Maps https://github.com/segmentio/analytics-android/blob/master/analytics/src/main/java/com/segment/analytics/ValueMap.java#L347-L359 and JSONObject https://github.com/segmentio/analytics-android/blob/master/analytics/src/main/java/com/segment/analytics/internal/Utils.java#L383-L394.